### PR TITLE
Basic interfaces needed for modularization

### PR DESCRIPTION
**Pull request by Neeharika-Sompalli**
Related to #4024 

- Adds state package interfaces needed for modularization
- This PR does not add complete implementation for the interfaces. It is needed to unblock the parallel work that can be done related to modularization
- Since the first task is focused to modify `SigRequirements` part of the code to use new interfaces , all the changes deal with immutable state.

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/base/state/InMemoryStateImpl.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    private final Instant lastModifiedTime;
+
+    InMemoryStateImpl(@NotNull final String stateKey, @NotNull final Instant lastModifiedTime) {
+        super(stateKey);
+        this.merkle = new MerkleMap<>();
+        this.lastModifiedTime = lastModifiedTime;
+    }
+
+    public InMemoryStateImpl(
+            @Nonnull String stateKey,
```
Comment (tinker-michaelj) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/base/state/InMemoryStateImpl.java:44]: We probably want to standardize on either `@Nonnull` _or_ `@NotNull`; and either always explicitly throw `NullPointerException` on a null argument; or always rely on `Objects.requireNonNull()`.

**Review by tinker-michaelj — DISMISSED** [review]
LGTM

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/base/state/InMemoryStateImpl.java (hunk 1 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/base/state/InMemoryStateImpl.java:43]: Agree. fixed to use `javax.annotation.Nonnull` annotation and `Objects.requireNonNull`

**Review by kimbor — COMMENTED** [review]
LGTM!

**Pull request merged**